std/range/package.d

```diff
@@ -8606,6 +8606,385 @@ if (isForwardRange!Source && hasLength!Source)
+
+                    static if (hasLength!Source)
+                    {
+                        import std.algorithm.comparison : min;
+                        const size_t endIndex = min(_endIndexes[upper-1] - _offset, source.length);
+                    }
+                    else
+                        const size_t endIndex = _endIndexes[upper-1] - _offset;
+
+                    return Chunks(_source[startIndex .. endIndex], _endIndexes[lower .. upper-1], startIndex);
+                }
+            }
+
+            // Used when slicing to avoid taking extra elements
+            private static struct UseMaxLength
+            {
+                this(bool use, size_t len=0)
+                {
+                    _use = use;
+                    _len = len;
+                }
+
+                void decrease() { if (_len != 0) _len--; }
+
+                @property bool empty() { return _len == 0; }
+
+                private:
+                    size_t _len;
+                    bool _use;
+                alias _use this;
+            }
+
+        private:
+            Source _source;
+            Indexes _endIndexes;
+            size_t _offset;
+            UseMaxLength _maxLen;
+            bool _isEmpty;
+        }
+        else
+        {
+            import std.typecons : RefCounted;
+
+            private static struct Chunk
+            {
+                private RefCounted!Impl impl;
+
+                @property bool empty() { return impl.curSizeLeft == 0 || impl.source.empty; }
+                @property auto front() { return impl.source.front; }
+                void popFront()
+                {
+                    impl.curSizeLeft--;
+                    impl.source.popFront();
+                }
+            }
+
+            private static struct Impl
+            {
+                private Source source;
+                private Indexes endIndexes;
+                private size_t curSizeLeft;
+            }
+
+            private RefCounted!Impl impl;
+            // for when there are extra elements in the source range
+            private bool tookLastChunk;
+            private size_t offset;
+
+            private this(Source source, Indexes endIndexes)
+            {
+                tookLastChunk = false;
+                offset = 0;
+                impl = RefCounted!Impl(source, endIndexes,
+                    source.empty ? 0 : (endIndexes.empty ? 0 : endIndexes.front));
+            }
+
+            @property bool empty() { return impl.source.empty || tookLastChunk; }
+            @property Chunk front()
+            {
+                return Chunk(impl);
+            }
+
+            void popFront()
+            {
+                if (impl.endIndexes.empty)
+                {
+                    tookLastChunk = true;
+                    return;
+                }
+
+                impl.curSizeLeft -= impl.source.popFrontN(impl.curSizeLeft) - offset;
```

Comment:
I think the subtraction of offset is not correct here. Doesn't `curSizeLeft` denote the amount left in the chunk?

Reply:
Actually, I can't see any reason to reassign to impl.curSizeLeft. If popFrontN succeeds in popping off N elements, it will be set to 0. If it doesn't pop that many off, then the source is empty, and curSizeLeft is irrelevant.